### PR TITLE
RABSW-1003: Create a common ResourceError struct

### PR DESCRIPTION
**Pull request by matthew-richerson**
This commit creates a struct that can be embedded in the status section
of a resource to provide error information. The struct contains:
 - UserMessage: Externally facing error message meant for users
 - DebugMessage: Internal error message meant to provide extra information
for debugging
 - Recoverable: Indication of whether the error is likely to be recoverable

This commit adds the ResourceError struct to the ClientMount resource and
makes the necessary changes to the controllers.

Signed-off-by: Matt Richerson <mattr@cray.com>

### CODE REVIEW

**Review by ajfloeder — APPROVED** [review]

File: api/v1alpha1/resource_error.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+			message = message + ": " + err.Error()
+		}
+	}
+
+	resourceError.DebugMessage = message
+
+	return resourceError
+}
+
+func (e *ResourceError) GetDebugMessage() string {
```
Comment (NateRoiger) [inline api/v1alpha1/resource_error.go:59]: Why do you need Getters & Setters if the members are public? Should they be private?

**Review by NateRoiger — APPROVED** [review]

File: api/v1alpha1/resource_error.go (hunk 1 above)
Comment (matthew-richerson) [reply api/v1alpha1/resource_error.go:59]: I just did that out of habit. I don't think you can have private fields in a struct that's part of a k8s resource. I can remove them.

File: mount-daemon/controllers/clientmount_controller.go [hunk 2]
```diff
@@ -125,15 +125,25 @@ func (r *ClientMountReconciler) Reconcile(ctx context.Context, req ctrl.Request)
 		return ctrl.Result{}, nil
 	}
 
+	clientMount.Status.Error = nil
```
Comment (roehrich-hpe) [inline mount-daemon/controllers/clientmount_controller.go:128]: You shouldn't have to do this.  Nil should be the zero value for a pointer in Go.

Comment (matthew-richerson) [reply mount-daemon/controllers/clientmount_controller.go:128]: clientMount.Status.Error may have been set in a previous reconcile() call.

**Pull request merged**